### ISSUE TITLE
Issues in usage of hexagon delegate in raspi(armv7) and hikey(aarch64).

### ISSUE THREAD

**Issue by pranathibl** [tensorflow/tensorflow#37154]
I was running label_image.cpp tflite example. I was modifying some of the default parameters and trying to make use of them. I am also trying to make use of hexagon delegate API in raspi, Hikey boards. 

I have few queries in the example.
1. In documentation, it is mentioned the Hexagon Delegate API works in Qualcomm specific processors such as snapdragon 660,835 etc.. If I have development board having Qualcomm mentioned processors with Linux os installed. Could I be able to make use of it ?

2. The Tflite model expects input mean and std deviation to be specified. 
**Is this input mean and std deviation setting varies from model to model?**
**How to know what mean and std deviation to be kept to get better classification results based on model ?**
For Quantized models, We are not using mean and std deviation?
**Why is this only required for float models but not Quantized models ?**

3. what does **allow_fp16** ideally does ? I tried to use this flag but even after setting to true, didn't notice any difference in memory , performance and even interpreter state. 

In documentation, It is mentioned as it does fp16  precision calculations even for fp32. But interpreter tensors are always created in fp32(but not fp16).
**What happens when we use this flag for tflite (float) models ?**
**What happens when we use this flag for Quantized(uint8) models ?**

4. What does **profiling and max_profiling_buffer_entries** do ?
Which value i should keep for max_profiling_buffer_entries ? why is it 1024 ?
I was able to return elapsed times even when i set profiling flag to false and verbose to true.

Please provide, sufficient details in each question to help us in proceeding further.

**Comment by freedomtan** [comment]
1. Nope, Hexagon is Qualcomm's DSP shipped with its SoCs. Neither Raspberry Pi nor HiKey boards use Qualcomm SoCs. If you run `label_image --hexagon_delegate 1 ...`, you are supposed to get something like `Hexagon acceleration is unsupported on this platform.invoked` and TFLite interpreter will be used to run your model.

2. Usually models are trained with normalized/preprocessed images, those steps include converting images from integer to floating point numbers because models are trained in floating point. Check [1] for an example of preprocessing.

3. `--allow_fp16` is mainly for allowing running fp32 models with fp16 via `NNAPI` [2]
4.  To see TFLite per op profiling results, build `label_image` with  `--cxxopt=-DTFLITE_PROFILING_ENABLED` and run it with `-p 1` 


[1] https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/inception_preprocessing.py
[2] https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1gab822719f98f0c92e5da3684cdaca6ba0

**Comment by pranathibl** [comment]
If I have board with Qualcomm soc and which was running on Linux OS, Can I be able to use Hexagon Delegate. Is it restricted to only Android/ Ios ?

**Comment by terryheo** [comment]
1. Yes. Hexagon delegate is only available on Android at the moment.
2. You just need to provide dataset. Calculation is done by quantizer. 
https://www.tensorflow.org/lite/performance/post_training_quantization#full_integer_quantization_of_weights_and_activations

**Comment by pranathibl** [comment]
Hi terryheo, 

I have few issues regarding 2 point mentioned 

1. I was using label_image.cpp example in TfLite example. There there was option to pass mean and std deviation ? Why is it default set to 127.5 .

2. Does the TfLite Model has the information of what mean and std deviation ? If Quantizer does the calculation based on data set, Why this information is not stored in TfLite Model ?

Why in the example, he is not using model to get mean and std deviation info ? Why is he expecting the user to pass ?

3. If I use the quantizer and provided data set, How would I know, what quantizer has  set the mean and standard deviation ?

Please provide me details on the above questions.

**Comment by freedomtan** [comment]
when training image classifiers in floating point, usually image RGB values in [0,255], are converted to floating point numbers in [0,1] or [-1,1]. For, x in [0, 255], -1 <= (x-127.5) / 127.5  <= 1.0.

**Comment by pranathibl** [comment]
If i use pretrained tflite models from tensorflow lite website . How to know what kind of conversion is done to RGB values ? Can I use mean = 127.5 and std = 127.5 for all those models.

**Comment by freedomtan** [comment]
As far as I know, you have to check how the model is trained. So far it seems the classification models at the [TFLite hosted models](https://www.tensorflow.org/lite/guide/hosted_models) page are trained with inception processing, but I am not sure if that be will always true.

**Comment by arrow53** [comment]
> 1. Yes. Hexagon delegate is only available on Android at the moment.

@terryheo I have a debian based system. Is there a way to use the snapdragon hexagon delegate with that? This builds fine on my dragonboard 820c and now I'm trying to accelerate it.
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/label_image

**Comment by terryheo** [comment]
@arrow53 no. It's only supported by Android at the moment.
To enable it for ARM Linux, you may need to update https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/make/Makefile to include
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/experimental/delegates/hexagon
which we haven't tested it.
